### PR TITLE
Add `nix env shell`

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

Start implementing #10504.

> This is part of an effort to bring more structure to the CLI subcommands.
> 
> `nix env` will be about the process environment.
> Future commands may include `nix env run` and `nix env print-env`.
> 
> It is also somewhat analogous to the [planned](https://github.com/NixOS/nix/issues/10504) `nix dev shell` (currently `nix develop`), which is less about environment variables, and more about running a development shell, which is a more powerful command, but also requires more setup.

Most of the code has just moved to `env.cc`.

# Context

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/nix/env.cc [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    std::string description() override
+    {
+        return "manipulate the process environment";
+    }
+
+    Category category() override { return catUtility; }
+};
+
+static auto rCmdEnv = registerCommand<CmdEnv>("env");
```
Comment (roberth) [inline src/nix/env.cc:20]: This is new and perfectly analogous to `CmdStore`.

File: src/nix/env.cc [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        setEnv("PATH", unixPathString.c_str());
+
+        Strings args;
+        for (auto & arg : command) args.push_back(arg);
+
+        runProgramInStore(store, UseLookupPath::Use, *command.begin(), args);
+    }
+};
+
+static auto rCmdShell = registerCommand2<CmdShell>({"env", "shell"});
```
Comment (roberth) [inline src/nix/env.cc:97]: Moved verbatim, except for the last line.

The last line has the canonical name, which appears in the help.
`nix shell --help` opens that page directly, and all the rest is also transparently aliased, without warning.

**Review by Ericson2314 — APPROVED** [review]

File: doc/manual/rl-next/nix-env-shell.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+---
+synopsis: "`nix env shell` is the new `nix shell`, and `nix shell` remains an accepted alias"
+issues: 10504
+prs: 10807
+---
+
+This is part of an effort to bring more structure to the CLI subcommands.
+
+`nix env` will be about the process environment.
+Future commands may include `nix env run` and `nix env print-env`.
```
Comment (Mic92) [inline doc/manual/rl-next/nix-env-shell.md:10]: We will need a deprecation phase of `nix print-dev-env` if we want to get rid off it.
It's used by *A LOT* of people: https://github.com/nix-community/nix-direnv/blob/1223f170d83efbe3a6f494f073b1d6b2d3dbae94/direnvrc#L300
https://github.com/direnv/direnv/blob/6a73cc2d7d13054829501df8756db04e79ff6a29/stdlib.sh#L1310

Comment (Mic92) [reply doc/manual/rl-next/nix-env-shell.md:10]: Ok. Looks you make aliases anyway. I would expect given that direnv is also part of any Linux distribution we will still have the current version in some redhat in 10 years time.

Comment (roberth) [reply doc/manual/rl-next/nix-env-shell.md:10]: Yes, these commands are not going away - not even deprecated.
Aliases are cheap to support, and `nix shell` is genuinely useful as a shorthand.
`print-dev-env` we might eventually hide or something, but that's for a later PR.

**Pull request merged**